src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -49,13 +50,48 @@ internal partial class RazorCustomMessageTarget
             return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion ?? -1);
         }
 
-        using var _ = ListPool<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>.GetPooledObject(out var requestTasks);
-
         semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
         var textBuffer = csharpDoc.Snapshot.TextBuffer;
 
         cancellationToken.ThrowIfCancellationRequested();
+        ReinvocationResponse<VSSemanticTokensResponse?>? result = null;
+        if (semanticTokensParams.UsePreciseRanges)
+        {
+            var languageServerName = RazorLSPConstants.RazorCSharpLanguageServerName;
+            var lspMethodName = RazorLSPConstants.RoslynSemanticTokenRangesEndpointName;
+            var newParams = new SemanticTokensRangesParams()
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                Ranges = semanticTokensParams.Ranges
+            };
 
+            using (var disposable = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId))
+            {
+                result = await _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangesParams, VSSemanticTokensResponse?>(
+                    textBuffer,
+                    lspMethodName,
+                    languageServerName,
+                    SupportsPreciseRanges,
+                    newParams,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // Only null if the server doesn't support the new endpoint
+            if (result != null)
+            {
+                if (result?.Response?.Data is null)
+                {
+                    _logger?.LogDebug("Made one semantic token request to Roslyn for {count} ranges but only got null result back, due to sync issues", semanticTokensParams.Ranges.Length);
+                    // Weren't able to re-invoke C# semantic tokens but we have to indicate it's due to out of sync by providing the old version
+                    return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion ?? -1);
+                }
+
+                _logger?.LogDebug("Made one semantic token requests to Roslyn for {count} ranges", semanticTokensParams.Ranges.Length);
+                return new ProvideSemanticTokensResponse(new int[][] { result?.Response?.Data! }, semanticTokensParams.RequiredHostDocumentVersion);
+            }
+        }
+
+        using var _ = ListPool<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>.GetPooledObject(out var requestTasks);
         foreach (var range in semanticTokensParams.Ranges)
```

Comment:
There will only ever be one range here now, correct? Can assert that and remove all of the Task.Run stuff, right?

Reply:
I thought about cleaning it up but it would fallback into the loop until the new endpoint is available.

As long as we're going to wait until the new roslyn endpoint is available then yes, I can assert there would only be one range available and no Task.Run would be needed.

This means the CI tests will fail too until the new endpoint is available

Reply:
Hmmmm CI is a problem. That will be a long wait.